### DISCUSSION TITLE
Omitting headers/page numbers on empty pages

### DISCUSSION THREAD

**Discussion by RAnders00** [typst/typst#3122] [Q&A] [answered]
I am working on a template where each chapter is supposed to begin on its own odd page:

```typst
// Place new chapters on a new odd page
#show heading.where(level: 1): it => {
  pagebreak(weak: true, to: "odd")
  it
}
```

and I have headings and page numbers configured:

```typst
#set page(header: ["Header contents!"], numbering: "1")
```

[View a complete example here](https://typst.app/project/r1AXtF7_D4lFIgmUNWhbWn)

![Example](https://github.com/typst/typst/assets/1629196/e26715d3-1390-455b-80e8-9371b441cd20)

Unfortunately, **the header and page numbers are repeated on the completely empty pages**, which I find undesireable. I've played around with locate and query to try and figure out whether the current page has elements, but I haven't gotten anything to work. Is there a way to achieve this?

**Comment by Julian-Wassmann** [comment]
Hi,

This is possible to do, but it took me some time to figure it out.

At first, I thought it would be the easiest way to replace `pagebreak(weak: true, to: "odd")` with an own implementation and just insert an empty page if needed. But I couldn't figure out how to insert pages and query the page counter without conflicting those two.

The second attemp wraps chapter headings with markers (using a `label`). One `chapter-end-marker` before the `pagebreak` and the heading, and one `chapter-start-marker` after the heading. We can than query the pages of the markers and use those the information to decide if a certain page is an "inserted" page. With inserted I mean the page wich is added and should not contain the header.

Take a look at those two examples:
- `chapter-end-marker` is at page `1`, `chapter-start-marker` is at page `2` => there is no inserted page in between
- `chapter-end-marker` is at page `1`, `chapter-start-marker` is at page `3` => there is an inserted page in between

The `page-is-inserted` function does this decision. It needs to be wrapped in a `locate` function call. You can also use this function elsewhere, for example to hide the page numbering as well.

```typst
// returns true if page at location is an inserted page and false otherwise
#let page-is-inserted(loc) = {
  let page = loc.page()
  let pairs = state("chapter-markers").at(loc)
  if pairs == none { return false }
  // page is inserted if surrounded by end- and start-marker for any chapter
  return pairs.any(((end-page, start-page)) => {
    page > end-page and page < start-page
  })
}

#let custom-header = {
  locate(loc => {
    if not page-is-inserted(loc) [Header contents!]
  })
}
#let custom-numbering(page, maxPage) = {
  locate(loc => {
    if not page-is-inserted(loc) [#page]
  })
}
#set page(header: custom-header, numbering: custom-numbering, paper: "a6")

// wraps heading with end-marker and start-marker 
#show heading.where(level: 1): it => [
  #[]<chapter-end-marker> // marker positioned before the pagebreak
  #pagebreak(to: "odd", weak: true)
  #it
  #[]<chapter-start-marker> // marker positioned after the pagebreak
]

// queries markers and updates state with pairwise pages for each chapter
#locate(loc => {
  let chapter-end-markers = query(<chapter-end-marker>, loc)
  let chapter-start-markers = query(<chapter-start-marker>, loc)
  let pairs = chapter-end-markers.enumerate().map(((index, chapter-end-marker)) => {
    let chapter-start-marker = chapter-start-markers.at(index)
    let end-page = chapter-end-marker.location().page()
    let start-page = chapter-start-marker.location().page()
    (end-page, start-page)
  })
  state("chapter-markers").update(pairs)
})

= Chapter 1

== Section A
== Section B
== Section C

= Chapter 2

== Section A
== Section B
== Section C
```

Reproduction URL: https://typst.app/project/rbX2FRMrJFIIGCpoh15wq4

I agree that this is a complicated thing to do and it would be nice to have this built-in to the `pagebreak` function. Or maybe a package for it (I'm working on one already but it's not done yet (contains other utilities too)).

**Reply by RAnders00** [reply]
Thank you! This is super helpful. I'll probably keep this as a separate file and import into my template, since it's quite a lot of code for what can be succinctly bundled in a function call. :)

**Comment by GamerBene19** [comment]
I've tried to use this approach, but sadly it breaks the page numbers in the outline.
As far as I can tell this is because the pagebreak makes it technically span over multiple pages. The outline uses the first page the heading appears on.
This can also be observed when you include `loc.page()` for debug purposes. The page numbers are sort of interwoven 
![image](https://github.com/typst/typst/assets/20821976/cfbdfccd-14d9-4c7a-a368-4e28ccb98e32)

**Reply by janekfleper** [reply]
Are you sure this is not just the difference between the value of `counter(page)` and the absolute page number `loc.page()`? If possible, could you provide an example that reproduces your issue?

**Reply by GamerBene19** [reply]
Yes. it appears to be related to the labels that are used here.
```
#set page(header: locate(loc => [Header is on page #loc.page()]))

#show heading.where(level: 1): it => locate(loc => [
  #[]<label>
  #pagebreak(to: "odd", weak: true)
  Heading is on page #loc.page() \
])

= L1 Heading
#locate(loc => [This text is on page #loc.page()])

= L1 Heading
#locate(loc => [This text is on page #loc.page()])
```
Produces 
![image](https://github.com/typst/typst/assets/20821976/9dc50804-e7d5-4d64-a383-6ac001128470)

If you remove/comment the line with the label, the heading location is as intended.

**Reply by janekfleper** [reply]
It looks like you can force the page counter to update by wrapping it in a context. I was not able to make this work within the scope of the `locate(loc => [])` function but I don't think you absolutely need to use this since you can just use `#here().page()` or `counter(page).display()` to get the current page. https://typst.app/project/rdMJcjjGmGQC25pkLndmTY

**Comment by julianmueller** [comment]
The solution that worked for me looks like this:

```typ
// Advances to the next odd page
#let pagebreak-to-odd = context {
  let n = here().page()
  if calc.odd(n) {
    page(header: [], footer: [])[]
  } else if calc.even(n) {
    v(100%, weak: true)
  }
}
```

so when I end on an odd page, I insert a blank page with no styling, and when I end on an even page, I simply add vertical spacing to fill up the page. By this I always start a new chapter on an odd page.

**Reply by laurmaedje** [reply]
Note that this will lead to layout convergence errors if used five times, because the `here().page()`'s are shifted by one in each compiler iteration.

**Reply by julianmueller** [reply]
Thanks for the info! Could you provide a snippet, that works as intended without causing this error?

**Reply by laurmaedje** [reply]
I've tried a bit and I don't think there's a clean solution at the moment. @Julian-Wassmann's answer still works as far as I can tell, except for the problem with the outline, but I don't really see a way to fix that.

**Reply by janekfleper** [reply]
Okay, I finally understood the issue with the outline and modified my solution that I already posted in the support thread on Discord. The page numbers in the outline are fixed now and the links will also point to the pages where the new chapters actually start.

So far I have not encountered any issues due to too many chapters or empty chapters. But if you do encounter some issues, please let me know. (I posted the modification of the footer in another comment below since the discussion was about removing both the header and the page number on empty pages.)

```typst
#let find-labels(name) = {
  return query(name).map(label => label.location().page())
}

#let page-header = context {
  let empty-pages = find-labels(<empty-page>)
  let new-chapters = find-labels(<new-chapter>)
  if new-chapters.len() > 0 {
    if new-chapters.contains(here().page()) [
      _a new chapter starts on this page_
      #return
    ]

    // get the index of the next <new-chapter> label
    let new-chapter-index = new-chapters.position(page => page > here().page())
    if new-chapter-index != none {
      let empty-page = empty-pages.at(new-chapter-index)
      if empty-page < here().page() [
        _this is an empty page to make the next chapter start on an odd page_
        #return
      ]
    }
  }

  [and this would be a regular header]
  line(length: 100%)
}

#show heading.where(level: 1): it => [
  #[] <empty-page>
  #pagebreak(to: "even", weak: true)
  #[] <new-chapter>
  #pagebreak(to: "odd", weak: true)
  #it.body
  #v(2em)
]

#show outline.entry.where(level: 1): it => {
  // reverse the results of the label queries to find the last <empty-page> label for the targeted page
  // the method array.position() will always return the first one...
  let empty-pages = find-labels(<empty-page>).rev()
  let new-chapters = query(<new-chapter>).rev()
  let empty-page-index = empty-pages.position(page => page == int(it.page.text))
  let new-chapter = new-chapters.at(empty-page-index)
  link(new-chapter.location())[#it.body #box(width: 1fr)[#it.fill] #new-chapter.location().page()]
}

#set page(header: page-header, numbering: "1")
```

You can also check out the project here https://typst.app/project/rmC_x90uIy0QcD-4lMcakP.

**Reply by julianmueller** [reply]
Okay, but you still have a number on your "blank" page, which to get rid of, was the reason for this endeavor.

**Reply by janekfleper** [reply]
Sure, this solution only shows the modification of the headers. But you can just apply the same principle to the footer. The function `page-header()` can differentiate between an `<empty-page>`, a `<new-chapter>` page and any other page.    

You can use this function to hide the page number on empty pages (but not on pages where a new chapter starts). It also alternates the position of the page number between left and right (for even and odd pages):
```typst
#let page-footer = context {
  // since the page breaks in chapter-heading() are inserted after the <empty-page> label,
  // the selector has to look "before" the current page to find the relevant label
  let empty-page-labels = query(selector(<empty-page>).before(here()))
  if empty-page-labels.len() > 0 {
    let empty-page = empty-page-labels.last().location().page()
    // look back at the most recent <new-chapter> label
    let new-chapter = query(selector(<new-chapter>).before(here())).last().location().page()
    // check that there is no <new-chapter> label on the current page
    if (new-chapter != here().page()) and (empty-page + 1 == here().page()) {
      return
    }
  }

  // left/right align the page number on even/odd pages
  let page-display = counter(page).display(here().page-numbering())
  return if calc.even(here().page()) { page-display } else { h(1fr) + page-display }
}
```

**Reply by julianmueller** [reply]
```typ
let page-display = counter(page).display(here().page-numbering())
//---------------------------------------^^^^^^^^^^^^^^^^^^^^^^^^
```
causes the Error: `Expected string, function, or auto, found none`

**Reply by janekfleper** [reply]
I added the `page-footer` to my example project here https://typst.app/project/rmC_x90uIy0QcD-4lMcakP.    

I could reproduce your error by omitting `#set page(numbering: "1")`. Please check if you have your page numbering set up correctly.

**Reply by julianmueller** [reply]
Now it works!

**Comment by jhalmen** [comment]
so i needed something similar and found this thread. The issue with the presented solutions is that the first content created in a `show heading` rule will be linked to automatically by typst. For the outline this can be worked around by redoing the links by hand and fixing the pagenumbers. It is however impossible to fix the bookmarks that can be selected in the pdf-viewer, they will always continue linking to the preceding `<chapter-end-marker>`.
These answers did however lead me to finding the solution that works correctly, without having to resort to fixing stuff by hand.
This is what I am using now. Instead of setting markers into the page it just updates a state variable, which then gets processed in every page's header:
```typst
#show heading.where(level: 1): it => {
  state("content.switch").update(false)
  pagebreak(weak: true, to:"odd")
  state("content.switch").update(true)

  it
}
#set page(
  header: context {
    let page = here().page()
    let is-start-chapter = query(heading.where(level:1))
      .map(it => it.location().page())
      .contains(page)
    if not state("content.switch", false).get() and not is-start-chapter {
      [ Empty Page ]
      return
    }
    state("content.pages", (0,)).update(it => {
          it.push(page)
          return it
        })
    if is-start-chapter {
      [ Start of chapter ]
    } else {
      [ There is content on page ]
    }
    line()
  },
  footer: context {
    let has-content = state("content.pages", (0,)).get()
      .contains(here().page())
    if has-content {
      align(center, counter(page).display())
    } else {
      [ Empty Page ]
    }
  }
)

```

**Reply by janekfleper** [reply]
This looks great! Thank you very much for posting your solution. I have to admit that I never checked the bookmarks in a PDF viewer since I was just playing around with the headers in a template in the web app...

I found that your solution only worked with version v0.12.0. When I tried version v0.11.1, the table of contents was not correct and the bookmarks were not working either. Do you know what actually changed between the versions?

**Reply by jhalmen** [reply]
yes, this was just fixed [recently](https://github.com/typst/typst/pull/4840#issue-2486763915)
> All locatable elements now emit a start and an end tag instead of just a single tag. This is primarily useful for future work on Tagged PDF, but it also allows us to correctly determine the exact position of an element in cases where we previously couldn't. In the case of show heading: it => pagebreak() + it, the heading's position will be after the pagebreak because neither visible content nor the matching end tag occurred before the pagebreak. Meanwhile, in metadata("hi") + pagebreak(), the position of the metadata will be before the pagebreak because it's end tag already appeared before the pagebreak. Notably, this means that the nesting of zero-sized elements is not necessarily preserved in the tag output (if the metadata was inside of the heading show rule, it remains before the pagebreak, but the heading migrates to after it).

As the comment mentions, this does not include the case where metadata is inserted before the pagebreak, hence the workaround with the state variables.

**Reply by maucejo** [reply]
@jhalmen Your solution works perfectly. I have to change some minor things to adapt it to my needs. Thank you for sharing this. However, it would be great if it was an option passed to `pagebreak` (even if I understand it will not the case in a near future).

**Reply by eduardz1** [reply]
Thanks to [nleanba on the forums](https://forum.typst.app/t/help-needed-in-making-layout-converge-within-5-attempts/5090/4?u=eduard) this can actually be simplified to just:

```typst
#show heading.where(level: 1, outlined: true): it => {
    pagebreak(weak: true)
    state("blank-page").update(true)
    pagebreak(to: "odd", weak: true)
    state("blank-page").update(false)
    it
}
#set page(footer: context {
    if not state("blank-page", false).get() {
      align(center, counter(page).display())
    }
 })
 ```
 
 resulting possibly in a bit less layout iterations

**Comment by lrsvmb** [comment, marked as answer]
I have found another way without using states or labels at all, producing an empty page before starting a new level 1 heading:

```typ
#let headings-on-odd-page(it) = {
  show heading.where(level: 1): it => {
    {
      set page(header: none, numbering: none)
      pagebreak(to: "odd")
    }
    it
  }
  it
}

#show heading.where(level: 1): it => // do your normal header styling here
#show: headings-on-odd-page
```

This works, because we restrict the scope of `set page(header: none, numbering: none)` to the scope inside the `headings-on-odd-page` show rule. Applying the show rule inside `headings-on-odd-page` will produce a header on an odd page, which is then matched by the show rule right before it (where you do your normal styling). This is then not matched again by our `headings-on-odd-page`, since that show rule has already been applied.

**Reply by eduardz1** [reply]
Thank you! This looks great!

**Reply by Ddystopia** [reply]
Super great, works like a charm!

**Reply by Splines** [reply]
Unfortunately, this doesn't seem to work anymore. We get a `page configuration is not allowed inside of containers` for this line:

```typ
set page(header: none, numbering: none)
```

**Reply by lrsvmb** [reply]
@Splines the code seems to work fine with Typst 0.14.2, which is the newest as of now. There are no errors logged, are you sure this error stems from my snippet?

**Reply by Splines** [reply]
@lrsvmb Sorry, I checked again and indeed this isn't a problem with your script, but rather on my end where I use this inside a custom template. Will have to further debug this. Anyways, thanks a lot for your solution!
